lib/charms/hpc_libs/v0/slurm_ops.py

```diff
@@ -197,9 +148,183 @@ def config_name(self) -> str:
+        ...
+
+    @abstractmethod
+    def service_for(self, service: ServiceType) -> ServiceManager:
+        """Get a `ServiceManager` for the specified `service`."""
+        ...
+
+    @abstractmethod
+    def configuration_manager_for(self, service: ServiceType) -> ConfigurationManager:
+        """Get a `ConfigurationManager` for the specified `service`."""
+        ...
+
+    @abstractmethod
+    def mungekey_manager(self) -> MungekeyManager:
+        """Get the `MungekeyManager` for munge."""
+        ...
+
+
+class MungeManager:
+    """Manage `munged` service operations."""
+
+    def __init__(self, ops_manager: SlurmPackageManager) -> None:
+        self.service = ops_manager.service_for(ServiceType.MUNGED)
+        self.config = ops_manager.configuration_manager_for(ServiceType.MUNGED)
+        self.key = ops_manager.mungekey_manager()
+
+
+class PrometheusExporterManager:
+    """Manage `slurm-prometheus-exporter` service operations."""
+
+    def __init__(self, ops_manager: SlurmPackageManager) -> None:
+        self.service = ops_manager.service_for(ServiceType.PROMETHEUS_EXPORTER)
+
+
+class SlurmManagerBase:
+    """Base manager for Slurm services."""
+
+    def __init__(self, service: ServiceType, ops_manager: SlurmPackageManager) -> None:
+        self.service = ops_manager.service_for(service)
+        self.config = ops_manager.configuration_manager_for(service)
+        self.munge = MungeManager(ops_manager)
+        self.exporter = PrometheusExporterManager(ops_manager)
+        self.ops = ops_manager
+
+    @property
+    def hostname(self) -> str:
+        """The hostname where this manager is running."""
+        return socket.gethostname().split(".")[0]
+
+
+class SnapPackageManager(SlurmPackageManager):
+    """Slurm ops manager that uses Snap as its package manager."""
+
+    def install(self) -> None:
+        """Install Slurm using the `slurm` snap."""
+        # FIXME: Pin slurm to the stable channel
+        _snap("install", "slurm", "--channel", "latest/candidate", "--classic")
+
+    def version(self) -> str:
+        """Get the current version of the `slurm` snap installed on the system."""
+        info = yaml.safe_load(_snap("info", "slurm"))
+        if (ver := info.get("installed")) is None:
+            raise SlurmOpsError("unable to retrive snap info. Ensure slurm is correctly installed")
+        return ver.split(maxsplit=1)[0]
+
+    def service_for(self, service: ServiceType) -> ServiceManager:
+        """Get a `ServiceManager` for the specified `service`."""
+        return _SnapServiceManager(service)
+
+    def configuration_manager_for(self, service: ServiceType) -> ConfigurationManager:
+        """Get a `ConfigurationManager` for the specified `service`."""
+        return _SnapConfigurationManager(service.config_name)
+
+    def mungekey_manager(self) -> MungekeyManager:
+        """Get the `MungekeyManager` for munge."""
+        return _SnapMungekeyManager()
```

Comment:
Tbh - I'm wondering if these should be properties instead? They're just providing proxied access to the objects that provide the specific methods for the given packaging format.

```suggestion
    @property
    def service_manager(self, service: ServiceType) -> ServiceManager:
        """Get a `ServiceManager` for the specified `service`."""
        return _SnapServiceManager(service)

    @property
    def config_manager(self, service: ServiceType) -> ConfigurationManager:
        """Get a `ConfigurationManager` for the specified `service`."""
        return _SnapConfigurationManager(service.config_name)

    @property
    def key_manager(self) -> MungekeyManager:
        """Get the `MungekeyManager` for munge."""
        return _SnapMungekeyManager()
```

Let me know your thoughts here :smile: